### PR TITLE
 Fix undefined files handling

### LINKED ISSUE #1472: bug: Generate Unit Test is broken
### Cody Version

main

### IDE Information

All

### Describe the bug

When trying to run test code generation I get:

```
2024-05-08 09:44:15,278 [ 234697]   WARN - #c.s.c.a.CodyAgentClient - Cody by Sourcegraph: █ telemetry-v2: recordEvent: cody.command.test/executed 
2024-05-08 09:44:16,740 [ 236159]   WARN - #c.s.c.a.CodyAgentClient - Cody by Sourcegraph: █ vscode.workspace.openTextDocument: unable to read non-file URI: untitled:/Users/pkukielka/Work/jetbrains-test2/src/test/kotlin/com/sourcegraph/cody/commands/CommandIdTest.kt
2024-05-08 09:44:16,742 [ 236161]   WARN - #c.s.c.a.CodyAgent - Error
2024-05-08 09:44:16,743 [ 236162]   WARN - #c.s.c.a.CodyAgent -     at FixupTask.set state (/Users/pkukielka/Work/sourcegraph/cody/vscode/src/non-stop/FixupTask.ts:95:25)
2024-05-08 09:44:16,743 [ 236162]   WARN - #c.s.c.a.CodyAgent -     at FixupController.setTaskState (/Users/pkukielka/Work/sourcegraph/cody/vscode/src/non-stop/FixupController.ts:1174:14)
2024-05-08 09:44:16,743 [ 236162]   WARN - #c.s.c.a.CodyAgent -     at FixupController.error (/Users/pkukielka/Work/sourcegraph/cody/vscode/src/non-stop/FixupController.ts:931:14)
2024-05-08 09:44:16,743 [ 236162]   WARN - #c.s.c.a.CodyAgent -     at EditProvider.handleError (/Users/pkukielka/Work/sourcegraph/cody/vscode/src/edit/provider.ts:243:32)
2024-05-08 09:44:16,743 [ 236162]   WARN - #c.s.c.a.CodyAgent -     at EditProvider.handleFileCreationResponse (/Users/pkukielka/Work/sourcegraph/cody/vscode/src/edit/provider.ts:336:22)
2024-05-08 09:44:16,885 [ 236304]   WARN - #c.s.c.a.CodyAgentClient - Cody by Sourcegraph: █ vscode.workspace.openTextDocument: unable to read non-file URI: untitled:/Users/pkukielka/Work/jetbrains-test2/src/test/kotlin/com/sourcegraph/cody/commands/CommandIdTest.kt
2024-05-08 09:44:16,889 [ 236308]   WARN - #c.s.c.a.CodyAgentClient - Cody by Sourcegraph: █ vscode.workspace.openTextDocument: unable to read non-file URI: untitled:/Users/pkukielka/Work/jetbrains-test2/src/test/kotlin/com/sourcegraph/cody/commands/CommandIdTest.kt
2024-05-08 09:44:16,893 [ 236312]   WARN - #c.s.c.a.CodyAgentClient - Cody by Sourcegraph: █ vscode.workspace.openTextDocument: unable to read non-file URI: untitled:/Users/pkukielka/Work/jetbrains-test2/src/test/kotlin/com/sourcegraph/cody/commands/CommandIdTest.kt
2024-05-08 09:44:16,900 [ 236319]   WARN - #c.s.c.a.CodyAgentClient - Cody by Sourcegraph: █ vscode.workspace.openTextDocument: unable to read non-file URI: untitled:/Users/pkukielka/Work/jetbrains-test2/src/test/kotlin/com/sourcegraph/cody/commands/CommandIdTest.kt
```

### Expected behavior

It should generate test code properly and not fail with the error.

### Additional context

It seems to be caused by a cody version bump, and perhaps the way it handles paths in some place.

### PR DESCRIPTION
**Pull request by pkukielka**
Fixes https://github.com/sourcegraph/jetbrains/issues/1472

## Changes

Added new client capability which is partial support for `undefined` files.
IntelliJ does not support in-memory files, but is able to create real files on the disk if given url is a valid path.

This is fixing `Generate Unit Test` command.

Cody PR: https://github.com/sourcegraph/cody/pull/4088

## Test plan

Manual tests using`Generate Unit Test` command.

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentService.kt [hunk 1]
```diff
@@ -69,11 +71,21 @@ class CodyAgentService(project: Project) : Disposable {
       agent.client.onTextDocumentShow = Function { params ->
         CodyEditorUtil.showDocument(
             project,
-            params.uri,
+            URI.create(params.uri).withScheme("file"),
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentService.kt:74]: so the scheme can still be `untitled` here but we are overwriting it anyway, right?

File: src/main/kotlin/com/sourcegraph/cody/edit/sessions/FixupSession.kt [hunk 2]
```diff
@@ -266,7 +263,7 @@ abstract class FixupSession(
 
     for (op in workspaceEditParams.operations) {
 
-      op.uri?.let { createAndSwitchFileIfNeeded(it) }
+      op.uri?.let { updateEditorIfNeeded(it) }
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/edit/sessions/FixupSession.kt:266]: so this is no longer a fixup session's responsibility to create the files... the other endpoint handles that now

**Comment by dominiccooney** [comment]
Just a heads up, this could interact with Cody Ignore which watches file:/// URLs. When the remote has an policy which could ignore anything, we will conservatively start blocking Cody in file:/// URLs which don't resolve to a remote repository.

File: src/main/kotlin/com/sourcegraph/cody/edit/sessions/FixupSession.kt (hunk 2 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/edit/sessions/FixupSession.kt:266]: Yep :) Now it is done how I should do it in the first place if I knew better :)

File: src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentService.kt (hunk 1 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentService.kt:74]: Yes, `untitled` is VSCode thing which leaks to use a little bit.
Those are really a files.

**Comment by pkukielka** [comment]
> Just a heads up, this could interact with Cody Ignore which watches file:/// URLs. When the remote has an policy which could ignore anything, we will conservatively start blocking Cody in file:/// URLs which don't resolve to a remote repository.

Thanks for pointing this out @dominiccooney .
I'm not 100% sure if I understands all possible interactions there.
This code is responsible for creating new files, when `Generate Unit Test` action will be executed.
Can such file creation be blocked on the Cody side even after the inline edit was already started for the source file?
If so, it would be great to add test for the agent to document corner cases of this behaviour.

**Review by mkondratek — APPROVED** [review]
